solr/solrj/src/java/org/apache/solr/common/util/CompressionUtil.java

```diff
@@ -0,0 +1,103 @@
+
+    private final static byte[] ZLIB_MAGIC = new byte[]{ 0x78, 0x1 };
+    private final static int COMPRESSED_SIZE_MAGIC_NUMBER = 2018370979;
+
+    /**
+     * Uses the hex magic number for zlib compression '78 01' to check if the bytes are compressed
+     * @param data - the bytes to check for compression
+     * @return true if the data is zlib compressed based on the first 2 bytes, false otherwise
+     */
+    public static boolean isCompressedBytes(byte[] data) {
+        if (data == null || data.length < 2) return false;
+        return ZLIB_MAGIC[0] == data[0] && ZLIB_MAGIC[1] == data[1];
+    }
+
+    /**
+     * Decompresses zlib compressed bytes, returning the uncompressed data as a byte[]
+     * @param data the input zlib compressed data to decompress
+     * @return the decompressed bytes
+     * @throws DataFormatException - The data is not zlib compressed data
+     */
+    public static byte[] decompressBytes(byte[] data) throws DataFormatException {
+        if (data == null) return null;
+        Inflater inflater = new Inflater();
+        inflater.setInput(data, 0, data.length);
+        // Attempt to get the decompressed size from trailing bytes, this will be present if compressed by Solr
+        int decompressedSize = ByteBuffer.wrap(data, data.length - 8, 4).getInt();
+        int xoredSize = ByteBuffer.wrap(data, data.length - 4, 4).getInt();
+        if ((decompressedSize ^ COMPRESSED_SIZE_MAGIC_NUMBER) != xoredSize) {
+            // Take best guess of decompressed size since it wasn't included in trailing bytes, assume a 5:1 ratio
+            decompressedSize = 5*data.length;
+        }
+        byte[] buf = new byte[decompressedSize];
+        int actualDecompressedSize = 0;
+        while (!inflater.finished()) {
+            if (actualDecompressedSize >= buf.length) {
+                buf = Arrays.copyOf(buf, (int)(buf.length*1.5));
+            }
+            actualDecompressedSize += inflater.inflate(buf, actualDecompressedSize, decompressedSize);
+        }
+        inflater.end();
+        if (buf.length != actualDecompressedSize) {
+            buf = Arrays.copyOf(buf, actualDecompressedSize);
+        }
+        return buf;
+    }
+
+    /**
+     * Compresses bytes into zlib compressed bytes using java native compression
+     * @param data the input uncompressed data to be compressed
+     * @return zlib compressed bytes
+     */
+    public static byte[] compressBytes(byte[] data) {
+        Deflater compressor = new Deflater(Deflater.BEST_SPEED);
+        compressor.setInput(data);
+        compressor.finish();
+        byte[] buf = new byte[data.length+8];
+        int compressedSize = 0;
+        while (!compressor.finished()) {
+            if (compressedSize >= buf.length) {
+                buf = Arrays.copyOf(buf, (int)(buf.length*1.5));
+            }
+            compressedSize += compressor.deflate(buf, compressedSize, buf.length - compressedSize);
+        }
+        compressor.end();
```

Comment:
nit: maybe use `try/finally` to ensure that `compressor.end()` is called in the `finally` block?